libmamba/src/specs/build_number_spec.cpp

```diff
@@ -0,0 +1,331 @@
+// Copyright (c) 2023, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#include <algorithm>
+#include <charconv>
+#include <stdexcept>
+
+#include <fmt/format.h>
+
+#include "mamba/specs/build_number_spec.hpp"
+#include "mamba/util/string.hpp"
+
+namespace mamba::specs
+{
+    /*********************************************************
+     *  BuildNumberPredicate BinaryOperators Implementation  *
+     *********************************************************/
+
+    auto BuildNumberPredicate::free_interval::operator()(const BuildNumber&, const BuildNumber&) const
+        -> bool
+    {
+        return true;
+    }
+
+    auto equal(BuildNumberPredicate::free_interval, BuildNumberPredicate::free_interval) -> bool
+    {
+        return true;
+    }
+
+    namespace
+    {
+        // Cannot specialize operator== in std namespace (UB) so we have to reinvent the wheel.
+
+        auto
+        equal(std::equal_to<BuildNumberPredicate::BuildNumber>, std::equal_to<BuildNumberPredicate::BuildNumber>)
+            -> bool
+        {
+            return true;
+        }
+
+        auto
+        equal(std::not_equal_to<BuildNumberPredicate::BuildNumber>, std::not_equal_to<BuildNumberPredicate::BuildNumber>)
+            -> bool
+        {
+            return true;
+        }
+
+        auto
+        equal(std::greater<BuildNumberPredicate::BuildNumber>, std::greater<BuildNumberPredicate::BuildNumber>)
+            -> bool
+        {
+            return true;
+        }
+
+        auto
+        equal(std::greater_equal<BuildNumberPredicate::BuildNumber>, std::greater_equal<BuildNumberPredicate::BuildNumber>)
+            -> bool
+        {
+            return true;
+        }
+
+        auto
+        equal(std::less<BuildNumberPredicate::BuildNumber>, std::less<BuildNumberPredicate::BuildNumber>)
+            -> bool
+        {
+            return true;
+        }
+
+        auto
+        equal(std::less_equal<BuildNumberPredicate::BuildNumber>, std::less_equal<BuildNumberPredicate::BuildNumber>)
+            -> bool
+        {
+            return true;
+        }
+
+        template <typename U, typename... T>
+        auto equal_with(const std::variant<T...>& lhs, const std::variant<T...>& rhs) -> bool
+        {
+            return std::holds_alternative<U>(lhs) && std::holds_alternative<U>(rhs)
+                   && equal(std::get<U>(lhs), std::get<U>(rhs));
+        }
+
+        template <typename... T>
+        auto equal(const std::variant<T...>& lhs, const std::variant<T...>& rhs) -> bool
+        {
+            return (equal_with<T>(lhs, rhs) || ...);
+        }
+    }
```

Comment:
I don't get the need for all this complexity. From what I understand (but I might be wrong), you want to ensure that the variant hold objects of the same type, which you can check with `lhs.index() == rhs.index()`

Reply:
From the definitions I read here, it looks like objects are equal as soon as they have the same type. Did I miss something?

Reply:
Ha I see what you mean. The code written does a proper variant comparison, but from the types included here (none have members), it is equivalent to only checking for type equality.
This is not what I wrote because this does not describe the proper intent. It may lead to subtle bugs should one modify/add a predicate.